google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionImpl.java

```diff
@@ -292,30 +293,38 @@ public void close() {
   }
 
   public ApiFuture<Void> closeAsync() {
-    if (!isClosed()) {
-      List<ApiFuture<Void>> futures = new ArrayList<>();
-      if (isBatchActive()) {
-        abortBatch();
-      }
-      if (isTransactionStarted()) {
+    synchronized (this) {
```

Comment:
This change looks bigger than it is. The actual change here is that:
1. The block is placed inside a `synchronized (this)` block.
2. The `futures.add(statementExecutor.submit(() -> null));` line is placed inside a try-catch block.